commands/pipelinecmd.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/GoogleContainerTools/kpt/internal/pipeline"
+
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+)
+
+// GetPipelineCommand returns a command that implements "kpt pipeline <run|...>" subcommands.
+func GetPipelineCommand(name string) *cobra.Command {
+	logger := makeLogger()
+	cmd := &cobra.Command{
+		Use: "pipeline",
+		// TODO(droot): wire docs with docs machinery
+		Short:   "Pipeline (coming soon)",
+		Long:    "Pipeline (coming soon)",
+		Example: "pipeline examples (coming soon)",
+		Aliases: []string{"p"}, // TODO(droot): kpt p run ?
+		RunE: func(cmd *cobra.Command, args []string) error {
+			defer logger.Sync()
+			h, err := cmd.Flags().GetBool("help")
+			if err != nil {
+				return err
+			}
+			if h {
+				return cmd.Help()
+			}
+			return cmd.Usage()
+		},
+	}
+
+	cmd.AddCommand(pipelineRunCmd(logger))
+	return cmd
+}
+
+// pipelineRunCmd returns an instance of cobra.Command that implements
+// "pipeline run" kpt subcommand.
+func pipelineRunCmd(logger *zap.Logger) *cobra.Command {
+	pkgPath := ""
+	c := &cobra.Command{
+		Use:     "run [DIR]",
+		Short:   "run",
+		Long:    "run",
+		Example: "run",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				// no pkg path specified, default to current working dir
+				wd, err := os.Getwd()
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "error getting current dir: %v \n", err)
+					return err
+				}
+				pkgPath = wd
+			} else {
+				// resolve and validate the provided path
+				pkgPath = args[0]
+			}
+			executor := pipeline.Executor{
+				PkgPath: pkgPath,
+				Logger:  logger,
+			}
+			err := executor.Execute()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "failed to run pipeline %v \n", err)
+				return err
+			}
+			return nil
+		},
+	}
+	return c
+}
+
+// makeLogger returns a logger.
+// Primary reason for having this is to help in debugging.
+// TODO(droot): reconcile this with kpt logger (if one exists).
+func makeLogger() *zap.Logger {
```

Comment:
What are the benefits of using the zap logging framework over klog?

Reply:
We used zap in kubebuilder related projects and were sort of happy with it (it's very flexible and doesn't come in the way), so I used it here. Are we using any logging in kpt yet ? Haven't used klog. If klog solves all the problems of glog, then may be that is a better alternative.